File: FE/mova/lib/app/models/series_model.dart
// To parse this JSON data, do
//
//     final seriesModel = seriesModelFromJson(jsonString);

import 'dart:convert';

SeriesModel seriesModelFromJson(String str) => SeriesModel.fromJson(json.decode(str));

String seriesModelToJson(SeriesModel data) => json.encode(data.toJson());

class SeriesModel {
  SeriesModel({
    this.status,
    this.pesan,
    this.data,
  });

  int? status;
  String? pesan;
  Data? data;

  factory SeriesModel.fromJson(Map<String, dynamic> json) => SeriesModel(
    status: json["status"],
    pesan: json["pesan"],
    data: Data.fromJson(json["data"]),
  );

  Map<String, dynamic> toJson() => {
    "status": status,
    "pesan": pesan,
    "data": data!.toJson(),
  };
}

class Data {
  Data({
    this.id,
    this.filmId,
    this.thumbnail,
    this.title,
    this.rating,
    this.category,
    this.genre,
    this.description,
    this.episode,
    this.release,
    this.region,
    this.createdAt,
    this.userFavorites,
  });

  int? id;
  int? filmId;
  String? thumbnail;
  String? title;
  double? rating;
  String? category;
  List<String>? genre;
  String? description;
  List<String>? episode;
  String? release;
  String? region;
  DateTime? createdAt;
  List<UserFavorite>? userFavorites;

  factory Data.fromJson(Map<String, dynamic> json) => Data(
    id: json["id"],
    filmId: json["film_id"],
    thumbnail: json["thumbnail"],
    title: json["title"],
    rating: json["rating"].toDouble(),
    category: json["category"],
    genre: List<String>.from(json["genre"].map((x) => x)),
    description: json["description"],
    episode: List<String>.from(json["episode"].map((x) => x)),
    release: json["release"],
    region: json["region"],
    createdAt: DateTime.parse(json["createdAt"]),
    userFavorites: List<UserFavorite>.from(json["user_favorites"].map((x) => UserFavorite.fromJson(x))),
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "film_id": filmId,
    "thumbnail": thumbnail,
    "title": title,
    "rating": rating,
    "category": category,
    "genre": List<dynamic>.from(genre!.map((x) => x)),
    "description": description,
    "episode": List<dynamic>.from(episode!.map((x) => x)),
    "release": release,
    "region": region,
    "createdAt": createdAt!.toIso8601String(),
    "user_favorites": List<dynamic>.from(userFavorites!.map((x) => x.toJson())),
  };
}

class UserFavorite {
  UserFavorite({
    this.id,
    this.filmId,
    this.userId,
    this.userUid,
  });

  int? id;
  int? filmId;
  int? userId;
  String? userUid;

  factory UserFavorite.fromJson(Map<String, dynamic> json) => UserFavorite(
    id: json["id"],
    filmId: json["film_id"],
    userId: json["user_id"],
    userUid: json["user_uid"],
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "film_id": filmId,
    "user_id": userId,
    "user_uid": userUid,
  };
}

File: BE/mova/models/series_model.go
package models

import (
	"io"
	"mime/multipart"
	"mova/db"
	"net/http"
	"os"

	"github.com/lib/pq"
)

func AddSeries(thumbnail *multipart.FileHeader, title string, rating float32, category string, genre []string, description string, episode []*multipart.FileHeader, release string, region string, createdAt string) (Response, error) {
	var series Series
	var film Film
	var res Response
	var err error
	var series_id int
	var film_id int
	var arrEpisodeUrl []string

	con := db.CreateCon()

	sqlStatement1 := "INSERT INTO film (thumbnail, title, rating, category, genre, release, region, createdAt, tipe, url) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) RETURNING id"
	sqlStatement2 := "INSERT INTO series (film_id, thumbnail, title, rating, category, genre, description, episode, release, region, createdAt) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11) RETURNING id"

	//thumbnail
	src, err := thumbnail.Open()
	if err != nil {
		return res, err
	}
	defer src.Close()

	thumbnailUrl := "thumbnail/" + thumbnail.Filename

	dst, err := os.Create(thumbnailUrl)
	if err != nil {
		return res, err
	}
	defer dst.Close()

	if _, err = io.Copy(dst, src); err != nil {
		return res, err
	}

	//episode
	for _, eps := range episode {
		//source
		src, err := eps.Open()
		if err != nil {
			return res, err
		}
		defer src.Close()

		epsUrl := "series/" + eps.Filename

		//destinanion
		dst, err := os.Create(epsUrl)
		if err != nil {
			return res, err
		}
		defer dst.Close()

		//copy
		if _, err = io.Copy(dst, src); err != nil {
			return res, err
		}

		arrEpisodeUrl = append(arrEpisodeUrl, epsUrl)
	}

	err = con.QueryRow(sqlStatement1, thumbnailUrl, title, rating, category, pq.Array(genre), release, region, createdAt, "series", arrEpisodeUrl[0]).Scan(&film_id)
	if err != nil {
		return res, err
	}

	film.Id = film_id
	film.Thumbnail = thumbnailUrl
	film.Title = title
	film.Rating = rating
	film.Category = category
	film.Genre = genre
	film.Release = release
	film.Region = region
	film.CreatedAt = createdAt
	film.Tipe = "series"
	film.Url = arrEpisodeUrl[0]

	err = con.QueryRow(sqlStatement2, film.Id, thumbnailUrl, title, rating, category, pq.Array(genre), description, pq.Array(arrEpisodeUrl), release, region, createdAt).Scan(&series_id)
	if err != nil {
		return res, err
	}

	series.Id = series_id
	series.FilmID = film.Id
	series.Thumbnail = thumbnailUrl
	series.Title = title
	series.Rating = rating
	series.Category = category
	series.Genre = genre
	series.Description = description
	series.Episode = arrEpisodeUrl
	series.Release = release
	series.Region = region
	series.CreatedAt = createdAt

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Add Series"
	res.Data = map[string]interface{}{
		"film":   film,
		"series": series,
	}

	return res, nil
}

func GetSeries(film_id string) (Response, error) {
	var userFavorite UserFavorite
	arrUserFavorite := []UserFavorite{}
	var rating float32
	var res Response
	var err error
	var series Series

	con := db.CreateCon()

	sqlStatement1 := "SELECT * FROM series WHERE film_id = ($1)"
	sqlStatement2 := "SELECT id, film_id, user_id, user_uid FROM favorite WHERE film_id = ($1)"
	sqlStatement3 := "SELECT AVG(rating) FROM rating WHERE film_id = ($1)"

	err = con.QueryRow(sqlStatement1, film_id).Scan(&series.Id, &series.FilmID, &series.Thumbnail, &series.Title, &series.Rating, &series.Category, pq.Array(&series.Genre), &series.Description, pq.Array(&series.Episode), &series.Release, &series.Region, &series.CreatedAt)
	if err != nil {
		return res, err
	}

	rowsFavorite, err := con.Query(sqlStatement2, film_id)
	if err != nil {
		return res, err
	}
	defer rowsFavorite.Close()

	for rowsFavorite.Next() {
		err = rowsFavorite.Scan(&userFavorite.Id, &userFavorite.FilmId, &userFavorite.UserId, &userFavorite.UserUID)
		if err != nil {
			return res, err
		}

		arrUserFavorite = append(arrUserFavorite, userFavorite)
	}
	series.UserFavorites = arrUserFavorite

	err = con.QueryRow(sqlStatement3, film_id).Scan(&rating)
	if err != nil {
		return res, err
	}
	series.Rating = rating

	res.Status = http.StatusOK
	res.Pesan = "Berhasil Get All Series"
	res.Data = series

	return res, nil
}
